docs: improve contributor guidelines (#33821)
cypress-io/cypress: CONTRIBUTING.md

### Pull Requests

**Before writing any code**, confirm that an open issue exists for the work and comment on it to let the team know you are interested in contributing. This gives us the opportunity to confirm we are willing to accept the contribution before you invest time in implementation. If no issue exists, [open one](https://github.com/cypress-io/cypress/issues/new/choose) first. For purely mechanical changes (typo fixes, documentation corrections), an issue is not required — include a brief explanation in the PR description instead.

- Break down pull requests into the smallest necessary parts to address the original issue or feature. This helps you get a timely review and helps the reviewer clearly understand which pieces of the code changes are relevant. A well-scoped PR:
  - Addresses **one concern** — a single bug, a single feature, a single refactor. Do not combine a functional change with an unrelated cleanup or style refactor, even if the cleanup is small.

>   - Is **readable in a single sitting**. As a rough guide, aim for logic diffs under ~50 lines. If your PR is growing beyond that, look for a natural seam to split it: extract a preparatory refactor PR first, then a feature PR on top.

Reviewer: I might say under 50 lines (excluding tests)

Author: Added '(excluding tests)' to the ~50 line guideline.

  - Avoids mixing concerns. If you notice an unrelated bug while fixing another, open a separate issue and PR rather than bundling the fix.
  - When in doubt, split. Reviewers can merge a short PR quickly; a large PR with mixed concerns slows everyone down.

> - When opening a PR for a specific issue already open, please name the branch you are working on using the convention `issue-[issue number]`. For example, if your PR fixes Issue #803, name your branch `issue-803`. If the PR is a larger issue, you can add more context like `issue-803-new-scrollable-area`. If there's not an associated open issue, **[create an issue](https://github.com/cypress-io/cypress/issues/new/choose)**.

Reviewer 2: This rule contradicts the typo/docs exception that was just added. Should the same exception apply here as well so contributors don't have conflicting instructions?

https://github.com/cypress-io/cypress/blob/4102a19a2ecdc4bfa28518dbdac739bc1ad3cdfa/CONTRIBUTING.md?plain=1#L515

More broadly, some of these guidelines are already difficult to follow because similar rules are stated differently in multiple places. Would it make sense to simplify or consolidate the guidance instead of continuing to add more? The document is already more than 700 lines and 7000 words long, which seems relevant even if line count isn’t a perfect metric.

Reviewer 3: @adamalston what is specifically contradictory here?

Reviewer 4: I think what @adamalston is referring to is "For purely mechanical changes (typo fixes, documentation corrections), an issue is not required" But this says to create an issue regardless of the type.

Author: Removed the 'If there's not an associated open issue, create an issue' tail from the branch-naming bullet — it contradicted the mechanical-changes carve-out. The paragraph above already covers when issues are and aren't required.

### Write Some Tests

> If you are adding a new feature or fixing a regression, ensure you add tests for it. PRs that introduce new behavior or fix a bug without corresponding tests are unlikely to be approved. The PR Tasks checklist in the pull request template requires confirming tests have been added or updated — this is a merge requirement, not a suggestion.

Reviewer 3: ```suggestion
Tests are required when developing new features or fixing regressions. PRs that introduce new behavior or fix a bug without corresponding tests are unlikely to be approved. The PR Tasks checklist in the pull request template requires confirming tests have been added or updated — this is a merge requirement, not a suggestion.
```

Author: Taken — updated to lead with the requirement as suggested.
